File: oblig4/administrator/java/Administrator.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet; 
import java.sql.SQLException;
import java.sql.Statement;

import java.util.LinkedList;
import java.util.List;
import java.util.Scanner;

public class Administrator {

    public static void main(String[] agrs) {

        String dbname = ""; // Input your UiO-username
        String user = ""; // Input your UiO-username + _priv
        String pwd = ""; // Input the password for the _priv-user you got in a mail
        // Connection details
        String connectionStr = 
            "user=" + user + "&" + 
            "port=5432&" +  
            "password=" + pwd + "";

        String host = "jdbc:postgresql://dbpg-ifi-kurs01.uio.no"; 
        String connectionURL = 
            host + "/" + dbname +
            "?sslmode=require&ssl=true&sslfactory=org.postgresql.ssl.NonValidatingFactory&" +
            connectionStr;

        try {
            // Load driver for PostgreSQL
            Class.forName("org.postgresql.Driver");
            // Create a connection to the database
            Connection connection = DriverManager.getConnection(connectionURL);

            int ch = 0;
            while (ch != 3) {
                System.out.println("-- ADMINISTRATOR --");
                System.out.println("Please choose an option:\n 1. Create bills\n 2. Insert new product\n 3. Exit");
                ch = getIntFromUser("Option: ", true);

                if (ch == 1) {
                    makeBills(connection);
                } else if (ch == 2) {
                    insertProduct(connection);
                }
            }
        } catch (SQLException|ClassNotFoundException ex) {
            System.err.println("Error encountered: " + ex.getMessage());
        }
    }

    private static void makeBills(Connection connection)  throws SQLException {
        // TODO: Oppg 2
    }


    private static void insertProduct(Connection connection) throws SQLException {
        // TODO: Oppg 3
    }

    /**
     * Utility method that gets an int as input from user
     * Prints the argument message before getting input
     * If second argument is true, the user does not need to give input and can leave
     * the field blank (resulting in a null)
     */
    private static Integer getIntFromUser(String message, boolean canBeBlank) {
        while (true) {
            String str = getStrFromUser(message);
            if (str.equals("") && canBeBlank) {
                return null;
            }
            try {
                return Integer.valueOf(str);
            } catch (NumberFormatException ex) {
                System.out.println("Please provide an integer or leave blank.");
            }
        }
    }

    /**
     * Utility method that gets a String as input from user
     * Prints the argument message before getting input
     */
    private static String getStrFromUser(String message) {
        Scanner s = new Scanner(System.in);
        System.out.print(message);
        return s.nextLine();
    }
}

File: oblig4/administrator/administrator.py
import psycopg2

# MERK: Må kjøres med Python 3!

# Login details for database user.
dbname = "martimn" #Set in your UiO-username
user = "martimn_priv" # Set in your priv-user (UiO-username + _priv)
pwd = "puPh3ouWah" # Set inn the password for the _priv-user you got in a mail

# Gather all connection info into one string.
connection = \
    "host='dbpg-ifi-kurs01.uio.no' " + \
    "dbname='" + dbname + "' " + \
    "user='" + user + "' " + \
    "port='5432' " + \
    "password='" + pwd + "'"


def administrator():
    conn = psycopg2.connect(connection)
    
    ch = 0
    while (ch != 3):
        print("-- ADMINISTRATOR --")
        print("Please choose an option:\n "
              "1. Create bills\n "
              "2. Insert new product\n "
              "3. Exit")
        ch = get_int_from_user("Option: ", True)

        if (ch == 1):
            make_bills(conn)
        elif (ch == 2):
            insert_product(conn)

    
def make_bills(conn):
    cur = conn.cursor()

    # Query the database for orders, join the relevant tables to create a tuple
    # with the precise values we need. I have chosen to list the totals for
    # each *order* in SQL, and do the summation per *user* in Python code
    # below.

    cur.execute(
        """
        SELECT uid,
               ws.users.name,
               ws.users.address,
               ws.orders.num * ws.products.price
         FROM ws.orders
              JOIN ws.products USING (pid)
              JOIN ws.users USING (uid)
        WHERE payed = 0;
        """
    )

    # I make use of dictionaries so store the values, it makes retrieval dead
    # simple when printing, below.

    recipients = {}
    addresses  = {}
    total_dues = {}

    for (uid, name, address, total) in cur.fetchall():
        if uid in total_dues:
            total_dues[uid] = total_dues[uid] + total
        else:
            recipients[uid] = name
            addresses[uid]  = address
            total_dues[uid] = total

    for (uid, total) in total_dues.items():
        if total == 0:
            continue
        print(f"---Bill---")
        print(f"Name: {recipients[uid]}")
        print(f"Address: {addresses[uid]}")
        print(f"Total due: {total}")
        print(f"")


def insert_product(conn):
    cur = conn.cursor()

    # Create a dictionary of valid category names to `cid`s beforehand. This
    # makes simplifies the validity check below.

    cids = {}
    cur.execute(
        """
        SELECT cid, name
          FROM ws.categories;
        """
    )
    for (cid, name) in cur.fetchall():
        cids[name] = cid

    print("-- Insert new product --")

    # Get name, category, price and description. Do an existence check for
    # `category`; only accept valid names. Also, make sure the `price` is an
    # actual float.

    name = input("Product name: ")

    while True:
        category = input("Category: ")
        if category in cids:
            category = cids[category]
            break
        else:
            print(f"Must be one of: {list(cids.keys())}")

    while True:
        try:
            price = float(input("Price: "))
            break
        except ValueError:
            print("Must be float!")

    description = input("Description: ")

    # Finally, insert and commit the values to database.

    cur.execute(
        """
        INSERT INTO ws.products (name, price, cid, description)
        VALUES (%s, %s, %s, %s);
        """,
        (name, price, category, description),
    )
    conn.commit()

    print(f"New product {name} inserted.")


def get_int_from_user(msg, needed):
    # Utility method that gets an int from the user with the first argument as
    # message Second argument is boolean, and if false allows user to not give
    # input, and will then return None.

    while True:
        numStr = input(msg)
        if (numStr == "" and not needed):
            return None;
        try:
            return int(numStr)
        except:
            print("Please provide an integer or leave blank.");


if __name__ == "__main__":
    administrator()
